src/transaction-validator.ts

```diff
@@ -19,16 +20,148 @@ export class TransactionValidator {
   validateTransaction(transaction: Transaction): ValidationResult {
     const errors: ValidationError[] = [];
 
-    // STUDENT ASSIGNMENT: Implement the validation logic above
-    // Remove this line and implement the actual validation
-    throw new Error('Transaction validation not implemented - this is your assignment!');
+    this.verifyUTXOExistance({ transaction, errors });
+
+    this.verifyNoDoubleSpend({ transaction, errors });
+
+    this.verifyNonNegativeOrZeroAmount({ transaction, errors });
+
+    if (
+      !this.hasError({ errors, code: VALIDATION_ERRORS.UTXO_NOT_FOUND }) &&
+      !this.hasError({ errors, code: VALIDATION_ERRORS.DOUBLE_SPENDING })
+    ) {
+      this.verifySignatures({ transaction, errors });
+      this.verifyBalance({ transaction, errors });
+    }
 
     return {
       valid: errors.length === 0,
       errors
     };
   }
 
+  private verifyUTXOExistance({
+    transaction,
+    errors
+  }: {
+    transaction: Transaction;
+    errors: ValidationError[];
+  }) {
+    transaction.inputs.forEach((input, i) => {
```

Comment:
Sin embargo, si es cierto que recorrés más de una vez el mismo array. Podría ser interesante implementar este método (y los demás) para UTXOs individuales, y después hacer un solo `for` en `validateTransaction`.